docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+ }
+ 
+ // SetGlobalSysVar implements GlobalVarAccessor.SetGlobalSysVar interface.
++// it is used for setting instance scope as well.
+ func (s *session) SetGlobalSysVar(name, value string) (err error) {
+        sv := variable.GetSysVar(name)
+        if sv == nil {
+@@ -1132,6 +1137,9 @@ func (s *session) SetGlobalSysVar(name, value string) (err error) {
+        if err = sv.SetGlobalFromHook(s.sessionVars, value, false); err != nil {
+                return err
+        }
++       if sv.HasInstanceScope() { // skip for INSTANCE scope
++               return nil
++       }
+        if sv.GlobalConfigName != "" {
+                domain.GetDomain(s).NotifyGlobalConfigChange(sv.GlobalConfigName, variable.OnOffToTrueFalse(value))
+        }
+@@ -1148,6 +1156,9 @@ func (s *session) SetGlobalSysVarOnly(name, value string) (err error) {
+        if err = sv.SetGlobalFromHook(s.sessionVars, value, true); err != nil {
+                return err
+        }
++       if !sv.HasInstanceScope() { // skip for INSTANCE scope
++               return nil
++       }
+        return s.replaceGlobalVariablesTableValue(context.TODO(), sv.Name, value)
+ }
+```
+
+A "non-native" `INSTANCE` variable can be changed to a native one as follows:
+
+```
++       {Scope: ScopeInstance, Name: TiDBGeneralLog, Value: BoolToOnOff(DefTiDBGeneralLog), Type: TypeBool, skipInit: true, SetGlobal: func(s *SessionVars, val string) error {
+                ProcessGeneralLog.Store(TiDBOptOn(val))
+                return nil
+-       }, GetSession: func(s *SessionVars) (string, error) {
++       }, GetGlobal: func(s *SessionVars) (string, error) {
+                return BoolToOnOff(ProcessGeneralLog.Load()), nil
+        }},
+```
+
+This introduces a compatibility issue, since users who have previously configured instance scope with `SET [SESSION] tidb_general_log = x` will receive an error because the scope is wrong. This can be fixed by adding a legacy mode to the sysvar framework which allows `INSTANCE` scoped variables to be set with either `SET GLOBAL` or `SET SESSION`:
+
+```sql
+SET GLOBAL tidb_enable_legacy_instance_scope = 1;
+```
+
+We can default to `TRUE` for now, but should revisit this in the future.
+
+### Stage 2: Mapping All Configuration File Settings to System Variables
+
+The second stage is to map all configuration file settings to system variable names. In MySQL any configuration file setting is also available as a system variable (even if it is read only), which makes it possible to get the configuration of the cluster with `SHOW GLOBAL VARIABLES`. Once this step is complete, it should also be possible to offer an `[instance]` section of the configuration file which accepts the system variable names (and finally consistent naming between the two systems). We can then modify the example configuration files to be based on the flat `[instance]` configuration:
+
+```
+# tidb.toml
+[instance]
+max_connections = 1234
+tidb_general_log = "/path/to/file"
```

Comment:
TiDB introduces ['CLUSTER_CONFIG'](https://docs.pingcap.com/tidb/stable/information-schema-cluster-config) since v4, so we need take it into consideration:

```
> SELECT * FROM INFORMATION_SCHEMA.CLUSTER_CONFIG;

...
tidb | ... | instance.max_connections = 1234
...
```
Imagine that it may be different with `select @@max_connections` because the instance variable is dynamic. This may be not a big issue but we need a clear clarification in the document.

Reply:
It will be important for Stage 3: refactoring, because the single source of truth is proposed to change. I can add it to the document for this stage. Prior to then, it is likely the `[instance]` version of the configuration will be an alias and update the heirachical version, but I can clarify that as well in stage 2.